src/repl_helper.rs

```diff
@@ -90,3 +99,148 @@ impl Validator for ReplHelper {
         self.validator.validate_while_typing()
     }
 }
+
+/// Mostly MatchingBracketHighlighter but with support for PartiQL bags. This
+/// allows, primarily, for multi-line input of bags.
+struct InputValidator;
+
+impl InputValidator {
+    fn new() -> InputValidator {
+        InputValidator {}
+    }
+}
+
+impl Validator for InputValidator {
+    fn validate(&self, ctx: &mut ValidationContext) -> RustylineResult<ValidationResult> {
+        Ok(validate_structure(ctx.input()))
+    }
+
+    fn validate_while_typing(&self) -> bool {
+        false
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum StructureCheck {
+    Single(char),
+    Repeat(char),
+}
+
+impl StructureCheck {
+    fn starts(c: char, next: Option<&char>) -> Option<StructureCheck> {
+        use StructureCheck::*;
+
+        match c {
+            '(' => Some(Single(')')),
+            '[' => Some(Single(']')),
+            '{' => Some(Single('}')),
+            '<' => {
+                if let Some('<') = next {
+                    Some(Repeat('>'))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn completes(c: char, next: Option<&char>) -> Option<StructureCheck> {
```

Comment:
Since this code is a mirror of L133 to L144 it would be nice to represent with a Map maybe?

Reply:
The more I worked with this code the less I liked what I was doing!